RegularGridTopology dimensions

epernod opened the pull request:
Small PR on GridTopology and RegularGridTopology link to issue: #163

- Allow to set RegularGridTopology in 1,2 or 3 dimension. But still check if the grid resolution is valid.
- Fix example of Tetrahedron FEM that was crashing if no tetrahedron is found in the scene.
- Add an example scene.
- Remove RegularGridTopology dependencies from MechanicalObject.

CHANGELOG:
  - GridTopology 
       - clean the code & factor the constructor
       - implement "flat" grids in 1 or 2 dimension by using setting grid resolution to "1" in the corresponding axis
  - RegularGridTopology clean the constructor's code & remove NDEBUG code. 
  - Adds RegularGridTopology_test
  - TetrahedronFEMForceField clean code
  - MechanicalObject removes code that was specific to the grid. 
  - Adds new examples scenes called: 
         examples/Components/topology/Hexa2TetraTopologicalMapping.scn
         examples/Components/topology/RegularGridTopology_dimension.scn
______________________________________________________
<!--- Please leave this at the end of your message -->
This PR: 
- [x] builds with SUCCESS for all platforms on the CI.
- [x] does not generate new warnings nor unit test failures.
- [x] does not break existing scenes.
- [x] does not break API compatibility.
- [x] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

epernod committed:
[SofaPhysicsAPI] FIX: delete, add helper init and remove dependencies to DrawTools in non GUI mode.

epernod committed:
Merge remote-tracking branch 'upstream/master'

epernod committed:
[GridTopology] ADD: enum to store the dimension of the current grid (to bu used in next commits)

epernod committed:
[GridTopology] FIX: changing grid resolution at init was not recomputing all data. Be careful order matter between regular grid and mecaObj...

epernod committed:
[FEM] FIX: TetrahedronFEMForceField has a pointer to the vector of tetra. This vector need to be set to something empty instead of NULL if no tetrahedron are in the scene.

epernod committed:
[GridTopology] FIX: allow to set grid resolution to 1 in every direction to have different dimension. If resolution is < 1, set to default value [ 2, 2, 2]

epernod committed:
[GridTopology] ADD: compute the list of triangles as well.

epernod committed:
[GridTopology] ADD: example scene with 3D, 2D and 1D grid

damienmarchal commented on SofaKernel/modules/SofaBaseTopology/GridTopology.cpp:
```diff
@@ -223,16 +238,34 @@ void GridTopology::setSize(int nx, int ny, int nz)
+
+    if (badDim)
     {
         msg_warning(this) << "The grid resolution: ["<< _n[0] << " ; " << _n[1] << " ; " << _n[2] <<
```
If the object inherit from Base you can now write 
msg_warning() instead of msg_warning(this)

damienmarchal commented on applications/projects/SofaPhysicsAPI/SofaPhysicsSimulation.cpp:
```diff
@@ -55,7 +56,13 @@ SofaPhysicsAPI::SofaPhysicsAPI(bool useGUI, int GUIFramerate)
 SofaPhysicsAPI::~SofaPhysicsAPI()
 {
-    delete impl;
+    std::cout << "SofaPhysicsAPI::~SofaPhysicsAPI()" << std::endl;
```
This looks like debugging so it should probably be removed or replaced with msg_info(...)

epernod replied:
totally debug... didn't see that commit end up in this PR ... Will remove that

damienmarchal reviewed (commented):
Thank @epernod  for PR, it fix existing problems, add examples and is good work. 
If you have time you can add a GridTopology_test somewhere to avoid regression and if you don't I will probably do one someday :)

epernod committed:
[GridTopology] FIX: remove some debug logs, change warning message.

epernod committed:
[GridTopology] FIX: bug in the constructors, the gridUpdate was not always created, thus topology was not computed.

epernod committed:
[GridTopology] ADD: test on RegularGridTopology testing the different creation of grid, sizes and topology creation.

epernod commented:
@damienmarchal I have added a RegularGridTopology_test. In fact I don't think GridTopology is used like this as a component. It is more a mother class for Regular/Cylindrical and sphereGridTopology (which each has tests now)

damienmarchal commented:
Thanks Eric for the fixes and the additional test.

Your answer about GridTopology rise other questions:
   - If the component is not supposed to be used without being inherited... shouldn't we remove it from the factory  ?
   - If a class is not a component does not mean we shouldn't test it. Actually my feeling is that base class (as they are shared) are very good entities to test :) 

EDIT: Maybe these questions should be for a new Issue and not to handle in this PR.

epernod commented:
In fact most of the methods I'm testing are directly inside the mother class. So I'm not sure a gridTopology_test is needed.
But to remove GridTopology from the factory is maybe possible. We need to investigate if someone is using it as a component.

damienmarchal commented:
Hi Erik, 

There is one test crash in  TetrahedronFEMForceField_test. Is this related to your changes ?

DM

epernod commented:
Should not but I will check.
Yes in fact my fix in TetrahedronFEMFF reveals that the test was doing nothing: This the log on the master branch:
```
[ RUN      ] TetrahedronDiffusionFEMForceField_test/0.extension
[ERROR]   [TetrahedronFEMForceField(fem)]  object must have a mesh topology. The component is inactivated.  To remove this error message please add a topology component to your scene.

```
Right now it crashes... will see if I understand enough the test to fix it.

guparan commented:
Hi @epernod 
Is this PR also linked to #162 ?

damienmarchal commented:
Hi @epernod 

Can I add more tests on this PR ? Because I think we should validate a bit more "negative/bad" case  to enforce the behavior of component when used improperly.

damienmarchal committed:
[SofaBaseTopology_test] CLEAN long namespace prefix with 'using'.

epernod commented:
help yourself.

epernod committed:
[FEM] FIX: pointer to the initial position need to be allocated to an empty vec and deleted only in case of bad topology. Should fix the test on tetrahedronFEM

epernod committed:
[GridTopology] UPDATE: remove the code related to the P0 Data of the regularGrid from the mechanicalObject reinit method. Check issue #42 for details.

damienmarchal committed:
Merge remote-tracking branch 'upstream/master' into regularGrid_dimensions

damienmarchal committed:
[SofaBaseTopology] ADD test on regular grid & clean constructors

First attempt to use the c++x11 constructor delegation to get rid
of the duplicated constructor in Sofa.

matthieu-nesme commented on SofaKernel/modules/SofaBaseTopology/GridTopology.cpp:
```diff
@@ -84,6 +85,23 @@ void GridTopology::GridUpdate::updateEdges()
+{
+    // base on quads
+    const SeqQuads& quads = topology->seqQuads.getValue();
+    SeqTriangles& triangles = *topology->seqTriangles.beginEdit();
```
prefer beginWriteOnly vs beginEdit when possible it saves an update.

epernod replied:
ok. Done.

matthieu-nesme commented on SofaKernel/modules/SofaBaseTopology/GridTopology.h:
```diff
@@ -195,6 +210,9 @@ using MeshTopology::getHexahedron;
     Data<bool> d_createTexCoords;
+
+    /// Enum storing the actual dimension of this grid
+    Grid_dimension m_gridDim;
```
what is the interest of storing this information?

epernod replied:
First idea was to use it from other components, like a tetrahedronFEM check that the grid is 3D to avoid crash.
But while testing I saw that we directly do test on the topology buffers.
So right now it is not used excepts in the test @damienmarchal added. But that the idea.

damienmarchal replied:
Up to my understanding, GridTopology have dimension once they are initialized...this is property of the topology that can be  recomputed from the nx,ny,nz each time we need it or stored. I have not opinion on that as long as there is function to retrieve the dimension.

damienmarchal committed:
Merge branch 'master' into regularGrid_dimensions

epernod committed:
[GridTopology] UPDATE: replace several beginEdit by beginWriteOnly

damienmarchal commented:
@matthieu-nesme  
Do you think we should remove the dimension ? 
Or is it ok to have this PR switched to the ready state ?

matthieu-nesme commented:
To me the notion of dimension is not well defined, keeping it as a function w/o Data would be better.
`unsigned getMaterialDimension() const` ?

untereiner commented:
I would suggest `unsigned getDimension() const` because the object should be considered as a topological space of a given dimension and not as a material since it does not have material properties yet.

epernod commented:
I agree, "Material" does not mean something in this base component. 
This parameter should only be informative, not a Data, there is no mechanism to change the grid resolution if you change this dimension.

epernod commented:
Finally do we have an agreement here? @matthieu-nesme @untereiner @damienmarchal 
Or do we remove that field?

damienmarchal commented:
getDimmension() sounds ok to me. 
Not needed to have it as a Data. 
It can be recomputed each time getDimmension is called, there is no need to store it.

damienmarchal commented:
@epernod, @matthieu-nesme  do you agreed so that we can merge soon ?

epernod commented:
It is ok for me. We are just speaking about an informative int or an enum. 
I can do the change (if any) as soon as their is a final decision.

damienmarchal commented:
@hugtalbot, you set this PR to ready but didn't dismissed @matthieu-nesme's "request for change", it  looks a bit ambiguous to me. 

Let's take it easy, I will dismissed the request for change and do the merge. If this is not what was intented don't hesitate to revert or submit a small fix to remove the Grid_dimmension as matthieu suggested.

damienmarchal commented:
[ci-build]

matthieu-nesme commented:
I still do not get the interest of storing `m_gridDim`. What is the limitation of a simple function? So you never have to change the value of this variable to keep it consistent.

If it was in a data you want to plug in a Data/Engine graph, it would indeed be necessary to store it, but it is not the case here.

damienmarchal commented:
Everyone agreed there is no interest in storing m_gridDim.  
9 days ago erik asked:
" Or do we remove that field?"  followed by "I can do the change (if any) as soon as their is a final decision."

He is waiting for an answer since 9 days.  
So here is the answer, "yes do the change and remove this field" :)

epernod committed:
[GridTopology] UPDATE: remove the m_gridDim member. getDimensions compute the current grid dimension using the d_n Data.

epernod commented:
yes we can change the test to remove` badDim` as now the dim is not computed at the same time.

No, `_n[i] = 1` works. `n = [1; 5; 5]` give a plan grid of 25 cells. -> Dim = 2

epernod committed:
[GridTopology] UPDATE: remove useless code in GridTopology::checkGridResolution()

matthieu-nesme commented:
@epernod If I am not wrong, if you have n=[2,2,2] your code will give dim=0 where it should give dim=3.
Only a test could check that ;)

epernod committed:
[GridTopology] FIX: bad computation of the grid dimension. Add new values in the test to cover that error.

epernod commented:
Indeed... my bad.
This Pr already add tests for the gridTopology but the Tests needed a test. 
Thanks @matthieu-nesme for the review. It should be good now.

damienmarchal commented:
[ci-build] 

EDIT: Thank you all to make things moving and progressing.

damienmarchal commented:
Hi Erik,

It seems that this PR introduced a new examples RegularGrid_dimmension that is failing because it use a component ```<EdgeSet/>``` that does not seems to be part of Sofa.

Can you provide a bit of details please.
